src/mjolnir/elevationbuilder.cc

```diff
@@ -34,22 +32,127 @@ constexpr double POSTING_INTERVAL = 60;
+        // mapped to a value between 0 to 15 for use in costing.
+        auto heights = sample->get_all(resampled);
+        auto grades = valhalla::skadi::weighted_grade(heights, interval);
+        if (length < kMinimumInterval) {
+          // Keep the default grades - but set the mean elevation
+          forward_grades = std::make_tuple(0.0, 0.0, 0.0, std::get<3>(grades));
+          reverse_grades = std::make_tuple(0.0, 0.0, 0.0, std::get<3>(grades));
+        } else {
+          // Set the forward grades. Reverse the path and compute the
+          // weighted grade in reverse direction.
+          forward_grades = grades;
+          std::reverse(heights.begin(), heights.end());
+          reverse_grades = valhalla::skadi::weighted_grade(heights, interval);
+        }
+      }
+
+      // Add elevation info to the geo attribute cache. TODO - add mean elevation.
+      float mean_elevation = std::get<3>(forward_grades);
+      uint32_t forward_grade = static_cast<uint32_t>(std::get<0>(forward_grades) * .6 + 6.5);
+      uint32_t reverse_grade = static_cast<uint32_t>(std::get<0>(reverse_grades) * .6 + 6.5);
+      auto inserted =
+          cache.insert({edge_info_offset,
+                        std::make_tuple(forward_grade, reverse_grade, std::get<1>(forward_grades),
+                                        std::get<2>(forward_grades), std::get<1>(reverse_grades),
+                                        std::get<2>(reverse_grades))});
+      found = inserted.first;
+
+      // Set the mean elevation on EdgeInfo
+      tilebuilder.set_mean_elevation(edge_info_offset, mean_elevation);
+    }
+
+    // Edge elevation information. If the edge is forward (with respect to the shape)
+    // use the first value, otherwise use the second.
+    bool forward = directededge.forward();
+    directededge.set_weighted_grade(forward ? std::get<0>(found->second)
+                                            : std::get<1>(found->second));
+    float max_up_slope = forward ? std::get<2>(found->second) : std::get<4>(found->second);
+    float max_down_slope = forward ? std::get<3>(found->second) : std::get<5>(found->second);
+    directededge.set_max_up_slope(max_up_slope);
+    directededge.set_max_down_slope(max_down_slope);
+  }
+
+  // Update the tile
+  tilebuilder.StoreTileData();
+
+  // Check if we need to clear the tile cache
+  if (graphreader.OverCommitted()) {
+    graphreader_lck.lock();
+    graphreader.Trim();
+    graphreader_lck.unlock();
+  }
+}
+
 /**
  * Adds elevation to a set of tiles. Each thread pulls a tile of the queue
  */
-void add_elevation(const boost::property_tree::ptree& pt,
-                   std::deque<GraphId>& tilequeue,
-                   std::mutex& lock,
-                   const std::unique_ptr<valhalla::skadi::sample>& sample,
-                   std::promise<uint32_t>& /*result*/) {
+void add_elevations_to_multiple_tiles(const boost::property_tree::ptree& pt,
```

Comment:
test coverage not required here as the whole method emerged from previously existing method, it was done just for reuse